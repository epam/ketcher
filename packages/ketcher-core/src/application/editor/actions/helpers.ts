import { Atom, Bond, Neighbor, StereoLabel, Struct } from 'domain/entities';
import { StereoValidator } from 'domain/helpers';

export function getStereoAtomsMap(
  struct: Struct,
  bonds: Array<Bond>,
  bond?: Bond,
) {
  const stereoAtomsMap = new Map();
  const correctAtomIds: Array<number> = [];

  bonds.forEach((bond: Bond | undefined) => {
    if (bond) {
      const beginNeighs: Array<Neighbor> | undefined = struct.atomGetNeighbors(
        bond.begin,
      );
      const endNeighs: Array<Neighbor> | undefined = struct.atomGetNeighbors(
        bond.end,
      );

      if (
        StereoValidator.isCorrectStereoCenter(
          bond,
          beginNeighs,
          endNeighs,
          struct,
        )
      ) {
        const stereoLabel = struct.atoms.get(bond.begin)?.stereoLabel;
        if (
          stereoLabel == null ||
          stereoAtomsMap.get(bond.begin)?.stereoLabel == null
        ) {
          stereoAtomsMap.set(bond.begin, {
            stereoParity: getStereoParity(bond.stereo),
            stereoLabel: stereoLabel || `${StereoLabel.Abs}`,
          });
        }
        correctAtomIds.push(bond.begin);
      } else {
        if (!correctAtomIds.includes(bond.begin)) {
          stereoAtomsMap.set(bond.begin, {
            stereoParity: Atom.PATTERN.STEREO_PARITY.NONE,
            stereoLabel: null,
          });
        }
        if (!correctAtomIds.includes(bond.end)) {
          stereoAtomsMap.set(bond.end, {
            stereoParity: Atom.PATTERN.STEREO_PARITY.NONE,
            stereoLabel: null,
          });
        }
      }
    }
  });

  // in case the stereo band is flipped, changed or removed
  // TODO the duplication of the code below should be fixed, mayby by function
  if (bond) {
    if (!correctAtomIds.includes(bond.begin)) {
      stereoAtomsMap.set(bond.begin, {
        stereoParity: Atom.PATTERN.STEREO_PARITY.NONE,
        stereoLabel: null,
      });
    }
    if (!correctAtomIds.includes(bond.end)) {
      stereoAtomsMap.set(bond.end, {
        stereoParity: Atom.PATTERN.STEREO_PARITY.NONE,
        stereoLabel: null,
      });
    }
  }

  return stereoAtomsMap;
}

function getStereoParity(stereo: number): number | null {
  let newAtomParity: number | null = null;
  switch (stereo) {
    case Bond.PATTERN.STEREO.UP:
      newAtomParity = Atom.PATTERN.STEREO_PARITY.ODD;
      break;
    case Bond.PATTERN.STEREO.EITHER:
      newAtomParity = Atom.PATTERN.STEREO_PARITY.EITHER;
      break;
    case Bond.PATTERN.STEREO.DOWN:
      newAtomParity = Atom.PATTERN.STEREO_PARITY.EVEN;
      break;
  }
  return newAtomParity;
}
