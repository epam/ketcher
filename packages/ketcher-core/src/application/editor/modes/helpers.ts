import {
  BackBoneSequenceNode,
  EmptySequenceNode,
  HydrogenBond,
  LinkerSequenceNode,
  SubChainNode,
} from 'domain/entities';
import { ITwoStrandedChainItem } from 'domain/entities/monomer-chains/ChainsCollection';

export function isNodeRestrictedForHydrogenBondCreation(
  node: SubChainNode | BackBoneSequenceNode | undefined,
) {
  return (
    !node ||
    node instanceof LinkerSequenceNode ||
    node instanceof BackBoneSequenceNode ||
    node instanceof EmptySequenceNode
  );
}

export function isTwoStrandedNodeRestrictedForHydrogenBondCreation(
  twoStrandedNode?: ITwoStrandedChainItem,
) {
  const senseNodeHydrogenBonds =
    twoStrandedNode?.senseNode?.monomers.reduce((acc, monomer) => {
      return acc.concat(monomer.hydrogenBonds);
    }, [] as HydrogenBond[]) || [];

  return Boolean(
    isNodeRestrictedForHydrogenBondCreation(twoStrandedNode?.senseNode) ||
      isNodeRestrictedForHydrogenBondCreation(twoStrandedNode?.antisenseNode) ||
      twoStrandedNode?.antisenseNode?.monomers.some((monomer) =>
        monomer.hydrogenBonds.some((hydrogenBond) => {
          return senseNodeHydrogenBonds.includes(hydrogenBond);
        }),
      ),
  );
}
