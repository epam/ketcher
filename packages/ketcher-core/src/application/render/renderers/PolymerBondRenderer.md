# Class: PolymerBondRenderer

## How Snake Bonds Algorithm Works

### Prerequisites:
1) Only attachment points ['R1', 'R2'] have snake bond mode;
2) Only Peptide and Chem monomers have snake bond mode;
3) If there is a second monomer, then Snake bond mode could be enabled only when 2 monomers are not on the same horizontal line which means the distance of their Y-axis model coordinate is more than 0.5.
4) There are still free attachment points for both monomers.
5) There is no more than 1 bond between the first and the second monomer.
6) Snake mode and Single Bond memu are enabled.

### Workig Flow

Events triggered from: `Bond.ts`

1. When mousedown a monomer, the snake bond will start to be created. 
Refer to `RenderersManager - addPolymerBond`
2. When mousemove, the Snake Bonds Algorithm will be constantly called to calculate the position of the snake bond and move it.  
Refer to `RenderersManager - movePolymerBond`
3. When mouseup a monomer, the snake bond creation will be finished. 
Refer to `RenderersManager - finishPolymerBondCreation`

### Algorithm for Drawing Snake Bonds:

It will first calculate the position of the second monomer relative to the first monomer, then draw the snake bond differently according to the result. 

1) ![Alt text](../../../../docs/data/snake-bond-top-right.png)  
If the second monomer is positioned to the **bottom-right** of the first one, it will verify if the current attachment point of the first monomer is `R1`, if not, then subsequent rules will be followed: 

* The snake bond will start with a horizontal line(with length of half of first monomer's width, plus LINE_FROM_MONOMER_LENGTH) originating from the first monomer's center;
* Then draw a left-bottom circular arc segment;
* Then draw a vertical line(with length of Y-axis distance between start and end position minus twice circular arc segment length);
* Then draw a top-right circular arc segment;
* Finally draw a horizontal line to connect the end position(with length of X-axis distance between start and end position, minus twice circular arc segment length, minus half of first monomer's width, minus LINE_FROM_MONOMER_LENGTH).    

  If this approach does not work, the start and end positions will be switched to recalculate and redraw the snake bond. This reverses the position calculation. For example, if the second monomer was initially at the bottom-right of the first one, after switching, the second monomer should be at the top-left of the first one. Consequently, the snake bond will be drawn differently. This aims to avoid a scanario in which both snake bonds connect both to the left side of a single monomer.

2) ![Alt text](../../../../docs/data/snake-bond-top-left.png) 
If the second monomer is positioned to the **top-left** of the first one, it will verify if the current attachment point of the first monomer is `R1`, if yes, the start and end positions will be switched in order to recalculate and redraw the snake bond. Otherwise, the following rules will apply: 

* The snake bond will start with a horizontal line(with length of half of first monomer's width, plus LINE_FROM_MONOMER_LENGTH) from the first monomer's center;
* Then draw a left-bottom circular arc segment;
* Then draw a vertical line(with length of first monomer's height);
* Then draw a top-left circular arc segment;
* Then draw a horizontal line(with length of X-axis distance between start and end position, plus twice LINE_FROM_MONOMER_LENGTH, plus first monomer's width);
* Then draw a right-up circular arc segment;
* Then draw a vertical line(with length of Y-axis distance between start and end position, minus first monomer's height);
* Then draw a bottom-right circular arc segment;
* Finally draw a horizontal line to connect the end position(with length of LINE_FROM_MONOMER_LENGTH, plus half of first monomer's width).

3) ![Alt text](../../../../docs/data/snake-bond-top-right.png)  
If the second monomer is to the **top-right** of the first one, it will also verify if the current attachment point of the first monomer is `R1`, if yes, the start and end positions will be switched in order to recalculate and redraw the snake bond. Otherwise, the following rules will apply: 

* The snake bond will start with a horizontal line(with length of half of first monomer's width, plus LINE_FROM_MONOMER_LENGTH) from the first monomer's center;
* Then draw a left-top circular arc segment;
* Then draw a vertical line(with length of Y-axis distance between start and end position, minus twice circular arc segment length, plus half of first monomer's height);
* Then draw a bottom-right circular arc segment;
* Finally draw a horizontal line to connect the end position(with length of X-axis distance between start and end position, minus twice circular arc segment length, minus half of first monomer's width and minus LINE_FROM_MONOMER_LENGTH).  

4) ![Alt text](../../../../docs/data/snake-bond-bottom-left.png) 
If the second monomer is to the **bottom-left** of the first one, it will also verify if the current attachment point of the first monomer is `R1`, if yes, the start and end positions will be switched in order to recalculate and redraw the snake bond. Otherwise, the following rules will apply: 

* The snake bond will start with a horizontal line(with length of half of first monomer's width, plus LINE_FROM_MONOMER_LENGTH) from the first monomer's center;
* Then draw a left-bottom circular arc segment;
* Then draw a vertical line(with length of VERTICAL_LINE_LENGTH);
* Then draw a top-left circular arc segment;
* Then draw a horizontal line(with length of X-axis distance between start and end position, plus twice LINE_FROM_MONOMER_LENGTH, plus first monomer's width);
* Then draw a right-bottom circular arc segment;
* Then draw a vertical line(with length of Y-axis distance between start and end position, minus 4 times circular arc segment length, minus VERTICAL_LINE_LENGTH);
* Then draw a top-right circular arc segment;
* Finally draw a horizontal line to connect the end position(with length of LINE_FROM_MONOMER_LENGTH, plus half of first monomer's width).

5) ![Alt text](../../../../docs/data/snake-bond-left.png) 
If the second monomer is to the **left** of the first one, it will also verify if the current attachment point of the first monomer is `R1`, if yes, the start and end positions will be switched in order to recalculate and redraw the snake bond. Otherwise, the following rules will apply: 

* The snake bond will start with a horizontal line(with length of half of first monomer's width, plus LINE_FROM_MONOMER_LENGTH) from the first monomer's center;
* Then draw a left-bottom circular arc segment;
* Then draw a vertical line(with length of Y-axis distance between start and end position, plus first monomer's height);
* Then draw a top-left circular arc segment;
* Then draw a horizontal line(with length of X-axis distance between start and end position, plus twice LINE_FROM_MONOMER_LENGTH, plus first monomer's width);
* Then draw a right-up circular arc segment;
* Then draw a vertical line(with length of first monomer's height);
* Then draw a bottom-right circular arc segment;
* Finally draw a horizontal line to connect the end position(with length of LINE_FROM_MONOMER_LENGTH, plus half of first monomer's width).

6) If none of the above approaches worked, a line will simply be drawn from the bond's starting position to its end.

