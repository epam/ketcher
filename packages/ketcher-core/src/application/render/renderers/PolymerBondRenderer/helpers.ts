export const SMOOTH_CORNER_SIZE = 5;

export const generateBend = (
  dx1: number,
  dy1: number,
  dx: number,
  dy: number,
): string => {
  return `q ${SMOOTH_CORNER_SIZE * dx1},${SMOOTH_CORNER_SIZE * dy1} ${
    SMOOTH_CORNER_SIZE * dx
  },${SMOOTH_CORNER_SIZE * dy} `;
};

export const CORNER_LENGTH = 4;
export const DOUBLE_CORNER_LENGTH = CORNER_LENGTH * 2;

export const generateCornerFromTopToRight = () => {
  return `c 0,4.418 3.582,${CORNER_LENGTH} ${CORNER_LENGTH},${CORNER_LENGTH}`;
};

export const generateCornerFromLeftToTop = () => {
  return `c 4.418,0 ${CORNER_LENGTH},-3.582 ${CORNER_LENGTH},-${CORNER_LENGTH}`;
};

export const generateCornerFromBottomToRight = () => {
  return `c 0,-4.418 3.582,-${CORNER_LENGTH} ${CORNER_LENGTH},-${CORNER_LENGTH}`;
};

export const generateCornerFromBottomToLeft = () => {
  return `c 0,-4.418 -3.582,-${CORNER_LENGTH} -${CORNER_LENGTH},-${CORNER_LENGTH}`;
};

export const generateCornerFromLeftToBottom = () => {
  return `c 4.418,0 ${CORNER_LENGTH},3.582 ${CORNER_LENGTH},${CORNER_LENGTH}`;
};

export const generateCornerFromTopToLeft = () => {
  return `c 0,4.418 -3.582,${CORNER_LENGTH} -${CORNER_LENGTH},${CORNER_LENGTH}`;
};

export const generateCornerFromRightToTop = () => {
  return `c -4.418,0 -${CORNER_LENGTH},-3.582 -${CORNER_LENGTH},-${CORNER_LENGTH}`;
};

export const generateCornerFromRightToBottom = () => {
  return `c -4.418,0 -${CORNER_LENGTH},3.582 -${CORNER_LENGTH},${CORNER_LENGTH}`;
};
