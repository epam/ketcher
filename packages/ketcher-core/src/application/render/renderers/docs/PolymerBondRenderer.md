# Class: [PolymerBondRenderer](../PolymerBondRenderer.ts)

## How Snake Bonds Algorithm Works

### Prerequisites:
1) Only attachment points ['R1', 'R2'] have snake bond mode;
2) Only Peptide and Chem monomers have snake bond mode;
3) If there is a second monomer, then Snake bond mode could be enabled only when 2 monomers are not on the same horizontal line which means the distance of their Y-axis model coordinate is more than 0.5.
4) There are still free attachment points for both monomers.
5) There is no more than 1 bond between the first and the second monomer.
6) Snake mode and Single Bond menu are enabled.

### Working Flow

Events triggered from: [Bond.ts](/packages/ketcher-core/src/application/editor/tools/Bond.ts)

1. When mousedown a monomer, the snake bond will start to be created. Refer [addPolymerBond](/packages/ketcher-core/src/application/render/renderers/RenderersManager.ts#L60)
2. When mousemove, the Snake Bonds Algorithm will be constantly called to calculate the position of the snake bond and move it.  
Refer to [movePolymerBond](/packages/ketcher-core/src/application/render/renderers/RenderersManager.ts#L67)
1. When mouseup a monomer, the snake bond creation will be finished. 
Refer to [finishPolymerBondCreation](/packages/ketcher-core/src/application/render/renderers/RenderersManager.ts#L92)

### Algorithm for Drawing Snake Bonds:

It will first calculate the position of the second monomer relative to the first monomer, then draw the snake bond differently according to the result. Possible scenarios:

1) ![Alt text](./snake-bond-bottom-right.png)  
If the second monomer is positioned to the **bottom-right** of the first one, it will verify if the current attachment point of the first monomer is `R1`:

    If yes: The start and end positions will be switched to recalculate and redraw a more natural snake bond. **For example, if the second monomer is to the bottom-right of the first one, but only the attachment point `R2` remains available, then the bond should connect first monomer's `R1` attachment point with the second monomer's `R2` attachment point. This is the exact scenario where the second monomer is positioned to the top-left of the first one if we switch the monomers.** This aims to avoid a scanario in which both snake bonds connect both to the left side of a single monomer.  

    If not: The bond will be drawn from the bond's starting position to its end, following the sequence depicted in the image.  

2) ![Alt text](./snake-bond-top-left.png)  
If the second monomer is positioned to the **top-left** of the first one, it will verify if the current attachment point of the first monomer is `R1`:

    If yes, the start and end positions will be switched in order to recalculate and redraw the snake bond. Otherwise, the bond will be drawn from the bond's starting position to its end, following the sequence depicted in the image. 

3) ![Alt text](./snake-bond-top-right.png)  
If the second monomer is to the **top-right** of the first one, it will also verify if the current attachment point of the first monomer is `R1`:

    If yes, the start and end positions will be switched in order to recalculate and redraw the snake bond. Otherwise, the bond will be drawn from the bond's starting position to its end, following the sequence depicted in the image. 

4) ![Alt text](./snake-bond-bottom-left.png)  
If the second monomer is to the **bottom-left** of the first one, it will also verify if the current attachment point of the first monomer is `R1`:

    If yes, the start and end positions will be switched in order to recalculate and redraw the snake bond. Otherwise, the bond will be drawn from the bond's starting position to its end, following the sequence depicted in the image. 

5) ![Alt text](./snake-bond-left.png)  
If the second monomer is to the **left** of the first one, it will also verify if the current attachment point of the first monomer is `R1`:

    If yes, the start and end positions will be switched in order to recalculate and redraw the snake bond. Otherwise, the bond will be drawn from the bond's starting position to its end, following the sequence depicted in the image. 

6) If none of the above approaches worked, a line will simply be drawn from the bond's starting position to its end.