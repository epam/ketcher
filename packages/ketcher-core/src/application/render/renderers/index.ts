export { ChemRenderer } from './ChemRenderer';
export { PeptideRenderer } from './PeptideRenderer';
export { BaseRenderer } from './BaseRenderer';
export { BaseMonomerRenderer } from './BaseMonomerRenderer';
export { PolymerBondRenderer } from './PolymerBondRenderer';
