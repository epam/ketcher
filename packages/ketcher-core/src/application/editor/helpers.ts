import { ketcherProvider } from 'application/utils';
import { BaseMode } from './modes/BaseMode';
import {
  DEFAULT_LAYOUT_MODE,
  HAS_CONTENT_LAYOUT_MODE,
  modesMap,
} from './modes';
import { KetSerializer } from 'domain/serializers';

export const initializeMode = (mode?: BaseMode): BaseMode => {
  if (mode) {
    return mode;
  }
  const ketcher = ketcherProvider.getKetcher();
  const isBlank = ketcher?.editor?.struct().isBlank();
  if (isBlank) {
    return new modesMap[DEFAULT_LAYOUT_MODE]();
  }
  return new modesMap[HAS_CONTENT_LAYOUT_MODE]();
};

export const parseMonomersLibrary = (monomersDataRaw: string | JSON) => {
  const monomersLibraryParsedJson =
    typeof monomersDataRaw === 'string'
      ? JSON.parse(monomersDataRaw)
      : monomersDataRaw;
  const serializer = new KetSerializer();
  const monomersLibrary = serializer.convertMonomersLibrary(
    monomersLibraryParsedJson,
  );

  return { monomersLibraryParsedJson, monomersLibrary };
};
