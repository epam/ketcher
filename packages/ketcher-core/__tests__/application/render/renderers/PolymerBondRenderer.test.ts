import { createPolymerEditorCanvas } from '../../../helpers/dom';
import { getFinishedPolymerBond } from '../../../mock-data';
import { PolymerBondRenderer } from 'application/render/renderers/PolymerBondRenderer';

describe('Polymer Bond Renderer', () => {
  it('should render bond', () => {
    const canvas: SVGSVGElement = createPolymerEditorCanvas();
    const polymerBond = getFinishedPolymerBond(0.25, 0.25, 2.25, 2.5);
    polymerBond.moveToLinkedMonomers();
    const polymerBondRenderer = new PolymerBondRenderer(polymerBond);
    global.SVGElement.prototype.getBBox = jest.fn();
    jest
      .spyOn(global.SVGElement.prototype, 'getBBox')
      .mockImplementation(() => ({ width: 30, height: 20 }));
    polymerBondRenderer.show();

    expect(canvas).toMatchSnapshot();
  });
});
