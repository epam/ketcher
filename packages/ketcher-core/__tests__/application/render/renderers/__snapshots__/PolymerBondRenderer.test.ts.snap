// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Polymer Bond Renderer should render bond 1`] = `
<svg
  height="500"
  id="polymer-editor-canvas"
  width="500"
>
  <defs>
    <symbol
      id="peptide"
    >
      <path
        class="monomer-body"
      />
    </symbol>
  </defs>
  <g
    class="drawn-structures"
  >
    <g
      pointer-events="stroke"
    >
      <line
        class="selection-area"
        pointer-events="stroke"
        stroke="#333333"
        stroke-width="1"
        x1="400"
        x2="3600"
        y1="400"
        y2="4000"
      />
      <line
        stroke="transparent"
        stroke-width="10"
        x1="400"
        x2="3600"
        y1="400"
        y2="4000"
      />
      <circle
        cursor="pointer"
        cx="3600"
        cy="4000"
        fill="transparent"
        pointer-events="none"
        r="1"
        stroke-width="10"
      />
    </g>
  </g>
</svg>
`;
